### ISSUE TITLE
WinSW.NETCore31.zip : how is it meant to be used ?

### ISSUE THREAD

**Issue by dreamflow** [winsw/winsw#416]
from https://github.com/kohsuke/winsw/releases
i've downloaded the newest WinSW.NETCore31.zip
and extracted it , but i didn't found documentation about it .

does someone know how it is meant to be used ?

**Comment by NextTurn** [comment]
> i've downloaded the newest WinSW.NETCore31.zip
and extracted it , but i didn't found documentation about it .

There is only one *.exe* in the zip. You can rename and use it as *WinSW.exe* is documented. But don't rename other files.

The actual difference between the *.exe* and the *.zip* is that the *.exe* is *self-contained* so it can run on a machine without any pre-installed runtime but it's larger (~ 30MB); the *.zip* is *runtime-dependent* so it can run only if the .NET Core runtime (think .NET Framework) is installed but it's small (< 1MB). And yes, the difference in extensions doesn't explain these.

@oleg-nenashev We could optionally make the self-contained one a single executable to make it portable and fit in the existing packaging scheme. Thoughts?

**Comment by dreamflow** [comment]
thank you for the detailed explanation ! now i understand .

i'm hoping for a release of a self-contained but not(!) single-executable
version of a .NET-Core based WinSW as a zip that contains the WinSW binaries
together with an embedded and trimmed .NET-Core-Runtime , something like
WinSW.NETCore31.SelfContained.win-x64.zip  .

the great atvantage would be easier installations of WinSW on target machines ,
because a .NET-Core-Runtime doesn't need to be installed or present .
and no more worries about .NET-Core-Runtime versions .

the available WinSW.NETCore31.x64.exe comes close , but extracts
when launched to a folder whose destination-path i cannot control .

in other words i'm hoping for a release similar to WinSW.NETCore31.x64.exe ,
but in zip-form , where the contents are directly runnable :
  - without an additional runtime-installation
  - without the WinSW-exe extracting on launch its runtime

**Comment by NextTurn** [comment]
> extracts when launched to a folder whose destination-path i cannot control

FYI you can change the extraction destination by setting the `DOTNET_BUNDLE_EXTRACT_BASE_DIR` environment variable.

**Comment by dreamflow** [comment]
thank you for the info about `DOTNET_BUNDLE_EXTRACT_BASE_DIR` !
this is very helpful to know for .NET single-executables .
i've tried it and it works !

but unfortunately it looks like it's not possible to configure
a windows-service so that the service (for example
WinSW.NETCore31.x64.exe) is launched with a new
env-var like `DOTNET_BUNDLE_EXTRACT_BASE_DIR` .

i hope that your PR https://github.com/kohsuke/winsw/pull/421 gets merged .
that would solve this issue and hopefully make
also many other WinSW users happy .

**Comment by oleg-nenashev** [comment]
In my review queue, I will try to get to it ASAP. Sorry for missing the pings

**Comment by dreamflow** [comment]
is there hope for the PR https://github.com/kohsuke/winsw/pull/421 to get merged ?

i'm hoping for a release of a self-contained (but not single-executable)
version of a .NET-Core based WinSW as a zip that contains the WinSW binaries
together with an embedded and trimmed .NET-Core-Runtime , because
in our internal windows-network i have no control about if or what versions
of .NET(Core)Runtimes are installed . unfortunately such installations become
more and more "political" .
as the software-developer , at least i have freedom about the runtimes and
libraries that i use for the software i develop , as long as those runtimes
and libraries can be used by just copying them into the folder of my
developed application .

a self-contained release of WinSW would enable me and hopefully others ,
to use WinSW on a broader scale , meaning on many more windows-PCs
within restricted internal windows-networks .

**Comment by oleg-nenashev** [comment]
Yes, I plan to work on merging it soon. Note that recent releases already include .NET Core binaries which pack WinSW and the required runtime in a single EXE file

**Comment by NextTurn** [comment]
@dreamflow You could try another kind of single executable (#568). This may be the only chance to get a 15MB executable... XML schema validation, YAML deserialization and command line parsing will all significantly contribute to the binary size.

(Hmm, I forgot to retain the CI artifacts. I can retrigger the build if you want to give it a try.)

**Comment by dreamflow** [comment]
@NextTurn thank you for the hint . great idea . but i need
also 32bit versions for our many Win7-32bit-PC's .

the size of WinSW is of no concern for my use-cases ,
because plenty of disk-space available .
even runtime- or library-sizes up to 1GB would be no problem .

what i'm looking for in practise , is runtimes or libraries that
i can just copy as a folder next to my app , without complicated
installations deep in to the windows-system and not scattered
arround , like settings saved in windows registry , or hidden
happenings or the worst : automatic updates .

**Comment by NextTurn** [comment]
FYI: Thanks to .NET 5, we'll be shipping a new kind of single executable that is not a self-extractor by the end of this year (not #568).

**Comment by dreamflow** [comment]
@NextTurn thank you for the info .
is your WinSW-new-kind-of-single-executable solution
based on ["Single-File Apps in .NET 5"](https://github.com/dotnet/runtime/issues/36590)  ?
it looks like , that non-.NET-assembliy-dll's like coreclr.dll, clrjit.dll, ...
won't be included when compiling a .NET-project into a single-file-exe
as described in the above link :
```
"Any other native runtime binaries (including runtime components like
coreclr and clrjit) are left unbundled beside the single-file app.
There is an optional setting to IncludeNativeLibrariesInSingleFile,
and extract them to disk on startup."
```
it looks like the ultimate goal of a
**single-file-executable-that-is-not-self-extractor**
is not on the .NET horizon so far .

it sounds more like it could be a
**most-files-in-1-executable-that-is-not-a-self-extractor-&-accompanied-by-some-native-DLL's**

in other words it looks like it still will be a runnable-single-FOLDER-solution ,
just with much less files , than the currently in .NET possible
runnable-single-FOLDER-solution .

i'm happy with both old and new runnable-single-FOLDER-solutions ,
as long as it's without self-extractor-magic or other hidden happenings .

**Comment by NextTurn** [comment]
Well, you're right about native libraries. Static linking of coreclr and clrjit will only be supported on Linux in .NET 5. Support for Windows may come in future versions, I think.

**Comment by NextTurn** [comment]
Found a note in the design doc:

> Ideally, we should use the SuperHost for self-contained single-file apps on Windows too. However, due to certain limitations in debugging experience and the ability to collect Watson dumps, etc., the CoreCLR libraries are left on disk beside the app.

FYI #421 has been merged into the v3 branch.
